sync-core/src/main/java/com/cloudant/sync/datastore/BasicDatastore.java

```diff
@@ -184,15 +236,12 @@ public BasicDocumentRevision getDocument(String id) {
         return getDocument(id, null);
     }
 
-    @Override
-    public BasicDocumentRevision getDocument(String id, String rev) {
-        Preconditions.checkState(this.isOpen(), "Database is closed");
-        Preconditions.checkArgument(!Strings.isNullOrEmpty(id), "DocumentRevisionTree id can not be empty");
+    private BasicDocumentRevision getDocumentInBlock(String id, String rev){
         Cursor cursor = null;
         try {
             String[] args = (rev == null) ? new String[]{id} : new String[]{id, rev};
             String sql = (rev == null) ? GET_DOCUMENT_CURRENT_REVISION : GET_DOCUMENT_GIVEN_REVISION;
-            cursor = this.sqlDb.rawQuery(sql, args);
+            cursor = sqlDb.rawQuery(sql, args);
             if (cursor.moveToFirst()) {
                 long sequence = cursor.getLong(3);
                 List<? extends Attachment> atts = attachmentManager.attachmentsForRevision(sequence);
```

Comment:
`attachmentsForRevision` calls the database for this datastore, and is able to because the database is a public property. It needs to not be, allowing the database object to escape makes it essentially impossible to maintain the serial nature of database calls we're trying to introduce.

That is, all access to the underlying SQL database needs to be mediated from a single class which ensures everything happens on a queue. It might be that a helper class is required given the database is accessed from various places.